keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.onSucceededPods.indexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `indexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
```

Comment:
How does "continue" intersect with CronJobs?  If I have a cronjob concurrencyPolicy of Forbid and the Job indicates Success, but is still running pods, what do I expect to happen?

Reply:
It's a great question. I believe that we can select either 2 actions like this:

1. The CronJob with the `concurrentPolicy=Forbid` creates a new Job based only on the job's result (completed or failed). It means that the CronJob creates a new Job regardless of remaining running pods by the `whenTargetAchived=continue`. Then, together with the `JobSuccessPolicy`, we introduce a new concurrentPolicy, `strictForbid`, that the CronJob with `concurrentPolicy=strictForbid` creates a new Job after all job's pods are finalized. The `strictForbid` means the CronJob doesn't create if there are remaining running pods.

2. The CronJob with the `concurrentPolicy=Forbid` doesn't create a new Job if there are remaining running pods.
After we find the use case that we want to create a new Job by the CronJob based only on the result of the job, we introduce the new field, `concurrentPolicyStrategy=FinalizedJob|FializedPods (default)`.

I think that option 2 would be better since the current CronJob with `cuncurrentPolicy=Forbid` creates a new Job based on the job that all pods implictly are finalized, and option 1 has the breaking changes.

@deads2k WDYT?

Reply:
I think both could be "breaking" depending on perspective.

1. I view the Forbid as, "as soon as a job result is available, the next instance can run".  In this case, delaying that start is  breaking change.
2. I view the Forbid as, "as long as any job pod is running, the next instance is blocked".  In this case, starting the next job is a breaking change.

The safer way to break is 2 as you also indicated above. However, given that 1 is a valid way to interpret the intent, the CronJob API should have a way to represent it.  I think this impacts both Forbid and Replace policies since Replace would logically need to halt the continuing pods and Forbid needs two different options: Forbid-until-all-pods-done (likely just Forbid) and ForbidUntilJobSuccessful

Reply:
@deads2k I think that we can avoid "breaking" change by modifying the following logic:

CronJob logic change to prevent breaking change:

- `cuncurrentPolicy=Forbid`: Replace `len(cronJob.Status.Active) > 0` with `len(cronJob.Status.Active) > 0 && len(runningPods) >0` in https://github.com/kubernetes/kubernetes/blob/7bb00356f06332e63a9f06acd42f1bdd8fc559d6/pkg/controller/cronjob/cronjob_controllerv2.go#L558. This means that the CronJob doesn't create "as long as any job pod is running".
- `cuncurrentPolicy=Replace`: Delete jobs that have running pods in https://github.com/kubernetes/kubernetes/blob/7bb00356f06332e63a9f06acd42f1bdd8fc559d6/pkg/controller/cronjob/cronjob_controllerv2.go#L574

> However, given that 1 is a valid way to interpret the intent, the CronJob API should have a way to represent it.

Also, I agree with this.

So, it might be better to apply the above `CronJob logic change to prevent breaking change` in the alpha stage, and then we add a new `concurrencyPolicy=ForbidUntilJobSuccessful` in the beta stage.

@deads2k @soltysh WDYT?

Reply:
> Quoting our API docs:
> 
> ```
> - "Forbid": forbids concurrent runs, skipping next run if previous run hasn't finished yet;
> ```
> 
> so going back to what I mentioned earlier. Succeeded is just an intermediary step, which allows you to act on job results earlier, w/o waiting for the rest. But the job is not finished. Finished is only when it reaches the final step which is succeeded or failed. We can't break that assumption about jobs.

@deads2k @soltysh It makes sense. By introducing `Succeeded` condition type separate from `Completed`, I guess that we can avoid any API and feature breaking, right?

Reply:
In the alpha stage, the CronJob concurrentPolicy works based on the Job `Completed` condition as before.
We will discuss the possibility of adding a new concurrentPolicy, "ForbidUntilJobSuccessful" when the graduation to beta.

Reply:
Done.